gtsam/geometry/SO4.cpp

```diff
@@ -0,0 +1,226 @@
+  }
+
+  // Build expX = exp(xi^)
+  Matrix4 expX;
+  using std::cos;
+  using std::sin;
+  const auto X2 = X * X;
+  const auto X3 = X2 * X;
+  double a2 = a * a, a3 = a2 * a, b2 = b * b, b3 = b2 * b;
+  if (a != 0 && b == 0) {
+    double c2 = (1 - cos(a)) / a2, c3 = (a - sin(a)) / a3;
+    return SO4(I_4x4 + X + c2 * X2 + c3 * X3);
+  } else if (a == b && b != 0) {
+    double sin_a = sin(a), cos_a = cos(a);
+    double c0 = (a * sin_a + 2 * cos_a) / 2,
+           c1 = (3 * sin_a - a * cos_a) / (2 * a), c2 = sin_a / (2 * a),
+           c3 = (sin_a - a * cos_a) / (2 * a3);
+    return SO4(c0 * I_4x4 + c1 * X + c2 * X2 + c3 * X3);
+  } else if (a != b) {
+    double sin_a = sin(a), cos_a = cos(a);
+    double sin_b = sin(b), cos_b = cos(b);
+    double c0 = (b2 * cos_a - a2 * cos_b) / (b2 - a2),
+           c1 = (b3 * sin_a - a3 * sin_b) / (a * b * (b2 - a2)),
+           c2 = (cos_a - cos_b) / (b2 - a2),
+           c3 = (b * sin_a - a * sin_b) / (a * b * (b2 - a2));
+    return SO4(c0 * I_4x4 + c1 * X + c2 * X2 + c3 * X3);
+  } else {
+    return SO4();
+  }
+}
+
+//******************************************************************************
+// local vectorize
+static SO4::VectorN2 vec4(const Matrix4& Q) {
+  return Eigen::Map<const SO4::VectorN2>(Q.data());
+}
+
+// so<4> generators
+static std::vector<Matrix4, Eigen::aligned_allocator<Matrix4> > G4(
+    {SO4::Hat(Vector6::Unit(0)), SO4::Hat(Vector6::Unit(1)),
+     SO4::Hat(Vector6::Unit(2)), SO4::Hat(Vector6::Unit(3)),
+     SO4::Hat(Vector6::Unit(4)), SO4::Hat(Vector6::Unit(5))});
+
+// vectorized generators
+static const Eigen::Matrix<double, 16, 6> P4 =
+    (Eigen::Matrix<double, 16, 6>() << vec4(G4[0]), vec4(G4[1]), vec4(G4[2]),
+     vec4(G4[3]), vec4(G4[4]), vec4(G4[5]))
+        .finished();
+
+//******************************************************************************
+template <>
+Matrix6 SO4::AdjointMap() const {
+  // Elaborate way of calculating the AdjointMap
+  // TODO(frank): find a closed form solution. In SO(3) is just R :-/
+  const Matrix4& Q = matrix_;
+  const Matrix4 Qt = Q.transpose();
+  Matrix6 A;
+  for (size_t i = 0; i < 6; i++) {
+    // Calculate column i of linear map for coeffcient of Gi
+    A.col(i) = SO4::Vee(Q * G4[i] * Qt);
+  }
+  return A;
+}
+
+//******************************************************************************
+template <>
+SO4::VectorN2 SO4::vec(OptionalJacobian<16, 6> H) const {
+  const Matrix& Q = matrix_;
+  if (H) {
+    // As Luca calculated, this is (I4 \oplus Q) * P4
+    *H << Q * P4.block<4, 6>(0, 0), Q * P4.block<4, 6>(4, 0),
+        Q * P4.block<4, 6>(8, 0), Q * P4.block<4, 6>(12, 0);
+  }
+  return gtsam::vec4(Q);
+}
+
+///******************************************************************************
+template <>
+SO4 SO4::ChartAtOrigin::Retract(const Vector6& xi, ChartJacobian H) {
+  if (H) throw std::runtime_error("SO4::ChartAtOrigin::Retract Jacobian");
```

Comment:
So SO4 does not have ChartJacobian?

Reply:
Retract Jacobian is complex and is only needed in specialized applications. Implemented fro SO(3) but not other variants.

Reply:
Copy that :)